### PR TITLE
Update PBS to generate bid ids for cache and events

### PR DESCRIPTION
**Pull request by BraslavskiyAndrey**

### CODE REVIEW

File: src/main/java/org/prebid/server/auction/BidResponseCreator.java [hunk 1]
```diff
@@ -846,14 +874,13 @@ private Bid toBid(BidderBid bidderBid,
...
-        final String eventBidId = ObjectUtils.defaultIfNull(generatedBidId, bid.getId());
+        final String generatedBidId = generatedBidIds.getGeneratedId(bidder, bid.getId(), bid.getImpid());
         final Video storedVideo = impIdToStoredVideo.get(bid.getImpid());
-        final Events events = createEvents(bidder, account, eventBidId, eventsContext);
+        final Events events = createEvents(bidder, account, generatedBidId, eventsContext);
         final ExtBidPrebidVideo extBidPrebidVideo = getExtBidPrebidVideo(bid.getExt());
 
         final ExtBidPrebid extBidPrebid = ExtBidPrebid.builder()
-                .bidid(generatedBidId)
+                .bidid(idGenerator.getType() == IdGeneratorType.uuid ? generatedBidId : null)
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/auction/BidResponseCreator.java:883]: Please add comment that
"Set generated value if auction.generate-bid-id:true"

File: src/main/java/org/prebid/server/auction/BidResponseCreator.java [hunk 2]
```diff
@@ -200,6 +202,11 @@ private static boolean isEmptyBidderResponses(List<BidderResponse> bidderRespons
         final Set<Bid> winningBids = newOrEmptySet(targeting);
         final Set<Bid> winningBidsByBidder = newOrEmptySet(targeting);
 
+        final GeneratedBidIds generatedBidIds = GeneratedBidIds.of(bidderResponses,
+                (ignored1, bid) -> idGenerator.getType() == IdGeneratorType.uuid
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/auction/BidResponseCreator.java:206]: *ignored

File: src/main/java/org/prebid/server/auction/BidResponseCreator.java [hunk 3]
```diff
@@ -441,22 +448,38 @@ private static boolean isValidForCaching(Bid bid) {
         return bid.getDealid() != null ? price.compareTo(BigDecimal.ZERO) >= 0 : price.compareTo(BigDecimal.ZERO) > 0;
     }
 
-    private Map<String, List<String>> getBidderAndVideoBidIdsToModify(List<BidderResponse> bidderResponses,
-                                                                      List<Imp> imps) {
+    private GeneratedBidIds getBidderAndVideoBidIdsToModify(
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/auction/BidResponseCreator.java:451]: Mb lets call it GeneratedVideoBidIds then ?

**Review by DGarbar — APPROVED** [review]
Current logic fix the mismatched beetwen events bid.id and generated in cache bid.id. 

But as you probably noticed `getBidderAndVideoBidIdsToModify` feels redundant. Look at `doCacheOpenrtb` method. GeneratedVideoBidId prepeared but real video bids are defined later.

So I created  https://github.com/prebid/prebid-server-java/pull/1100 which will prepare special BidInfo object wich will contain already prepeared information. Please take a look if you interested.

So for now, as this feature is urgent, LGTM.

File: src/main/java/org/prebid/server/auction/model/GeneratedBidIds.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    /**
+     * Returns generated id for bid identified by bidder, bidId and impId parameters.
+     */
+    public String getGeneratedId(String bidder, String bidId, String impId) {
+        if (bidderToBidIds.containsKey(bidder)) {
+            return bidderToBidIds.get(bidder).get(makeUniqueBidId(bidId, impId));
+        } else {
+            return null;
```
Comment (rpanchyk) [inline src/main/java/org/prebid/server/auction/model/GeneratedBidIds.java:74]: Seems we won't have case when bidder key is absent.
Not critical, just to think.

**Review by rpanchyk — CHANGES_REQUESTED** [review]

**Review by rpanchyk — APPROVED** [review]

**Pull request merged**